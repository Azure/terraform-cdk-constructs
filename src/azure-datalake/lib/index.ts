export * from "./filesystem";
export * from "./path";
export * from "./datalake";
