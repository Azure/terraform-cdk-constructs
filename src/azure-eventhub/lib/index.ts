export * from "./authorization";
export * from "./eventhubcluster";
export * from "./consumer";
export * from "./instance";
export * from "./kusto-connection";
export * from "./namespace";
