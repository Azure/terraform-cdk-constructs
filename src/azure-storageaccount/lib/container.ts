import {
  StorageBlob,
  StorageBlobConfig,
} from "@cdktf/provider-azurerm/lib/storage-blob";
import {
  StorageContainer,
  StorageContainerConfig,
} from "@cdktf/provider-azurerm/lib/storage-container";
import { Construct } from "constructs";

export class Container extends Construct {
  public readonly name: string;
  private readonly blobs: Map<string, Blob>;
  private readonly storageAccountName: string;
  private readonly storageContainerName: string;

  /**
   * Represents an Azure Storage Container within a specific Azure Storage Account.
   *
   * This class is designed for the creation and management of an Azure Storage Container, which serves as a unit of storage
   * that houses data objects, known as blobs. Containers are analogous to directories in a file system, and are used to organize
   * sets of blobs within a storage account. This class allows for granular control over blob storage, providing functionalities
   * such as setting access levels, managing metadata, and implementing security measures like encryption scopes.
   *
   * @param scope - The scope in which to define this construct, typically a reference to the Cloud Development Kit (CDK) stack.
   * @param id - The unique identifier for this instance of the container.
   * @param props - Configuration properties for the Azure Storage Container, based on the StorageContainerConfig interface. These include:
   *                - `name`: The name of the storage container, which must be unique within the associated storage account.
   *                - `storageAccountName`: The name of the storage account where this container is located.
   *                - `containerAccessType`: Defines the level of public access to the container. Options include 'container',
   *                  'blob', or 'none', controlling how the blobs within the container can be accessed.
   *                - `defaultEncryptionScope`: Specifies the default encryption scope for the container, enhancing data security.
   *                - `encryptionScopeOverrideEnabled`: Allows or prevents overriding the encryption scope on a per-blob basis.
   *                - `metadata`: Key-value pairs that provide additional information about the container, which can be used for
   *                  organizational, search, or billing purposes.
   *                - `timeouts`: Specifies custom timeout settings for CRUD operations on the container, providing control over
   *                  operation durations and retries.
   *
   * Example usage:
   * ```typescript
   * const storageContainer = new Container(this, 'MyContainer', {
   *   name: 'mycontainer',
   *   storageAccountName: 'mystorageaccount',
   *   containerAccessType: 'blob',
   *   defaultEncryptionScope: 'myEncryptionScope',
   *   encryptionScopeOverrideEnabled: true,
   *   metadata: {
   *     department: 'Finance'
   *   }
   * });
   * ```
   * This instantiation sets up a storage container with specified properties including access types and metadata. It is ready
   * to house blobs and provides methods to manage these blobs effectively.
   */
  constructor(scope: Construct, id: string, props: StorageContainerConfig) {
    super(scope, id);

    this.blobs = new Map<string, Blob>();

    // Create a storage container
    const container = new StorageContainer(this, "container", {
      name: props.name,
      storageAccountName: props.storageAccountName,
      containerAccessType: props.containerAccessType,
      metadata: props.metadata,
    });

    this.name = props.name;
    this.storageAccountName = props.storageAccountName;
    this.storageContainerName = container.name;
  }

  /**
   * Adds a blob to this Azure Storage Container.
   *
   * This method facilitates the addition of a blob to an Azure Storage Container managed by this class. It handles the creation and
   * configuration of the blob, including setting its type, source content, and metadata. This is useful for uploading various types
   * of unstructured data, such as images, videos, documents, or other binary files, into a cloud-based storage solution.
   *
   * @param blobName - The name of the blob to be added, which will be used as the blob's unique identifier within the container.
   * @param filePath - The file path or URL for the source of the blob's content. This specifies the location of the file to be uploaded.
   * @param props - Optional configuration properties for the blob, which include:
   *                - `type`: The type of the blob (e.g., 'Block', 'Append', 'Page'). Default is 'Block'.
   *                - `contentType`: The MIME type of the blob's content, such as 'application/octet-stream' for binary data. This helps browsers
   *                  and other clients handle the file correctly when it's downloaded or accessed.
   *                - `metadata`: A dictionary of key-value pairs to store as metadata with the blob. Metadata is typically used to store additional
   *                  details about the blob, such as tags, descriptions, or other attributes.
   *
   * @returns The newly created Blob object, which represents the blob added to the storage container.
   *
   * Example usage:
   * ```typescript
   * const storageBlob = storageContainer.addBlob('exampleBlob', './path/to/local/file.txt', {
   *   type: 'Block',
   *   contentType: 'text/plain',
   *   metadata: { customKey: 'customValue' }
   * });
   * ```
   * In this example, a new blob named 'exampleBlob' is added to the storage container. The content of the blob is sourced
   * from a local file specified by `filePath`. The blob is configured as a 'Block' type with 'text/plain' content type and
   * custom metadata. The method returns the blob instance for further use or reference.
   */
  addBlob(blobName: string, filePath: string, props?: StorageBlobConfig): Blob {
    const newStorageBlob = new Blob(this, blobName, {
      name: blobName,
      storageAccountName: this.storageAccountName,
      storageContainerName: this.storageContainerName,
      type: props?.type || "Block",
      source: filePath,
      contentType: props?.contentType || "application/octet-stream",
      metadata: props?.metadata || {},
    });
    this.blobs.set(blobName, newStorageBlob);
    return newStorageBlob;
  }
}

export class Blob extends Construct {
  readonly name: string;

  /**
   * Represents a blob within an Azure Storage Container.
   *
   * This class is responsible for the creation and management of a blob in an Azure Storage Container. Blobs are unstructured
   * data objects, which can include files like images, documents, videos, or any other file type. The Blob class provides a way
   * to manage these files in the cloud, allowing for scalable, durable, and accessible data storage. This class supports various
   * blob types such as block blobs for text and binary data, append blobs for log files, and page blobs for large volumes of
   * random access data.
   *
   * @param scope - The scope in which to define this construct, typically representing the Cloud Development Kit (CDK) stack.
   * @param id - The unique identifier for this instance of the blob.
   * @param props - Configuration properties for the Azure Storage Blob. These properties may include:
   *                - `name`: The name of the blob, which must be unique within the container.
   *                - `storageAccountName`: The name of the storage account in which the blob is stored.
   *                - `storageContainerName`: The name of the storage container in which the blob resides.
   *                - `type`: The type of the blob (e.g., 'Block', 'Append', 'Page').
   *                - `source`: The source of the blob's content, which could be a path to a file or a URL.
   *                - `contentType`: The MIME type of the blob's content, such as 'application/octet-stream' for binary data.
   *                - `metadata`: A dictionary of strings that represents metadata to associate with the blob.
   *                - `accessTier`: The data access tier, affecting storage costs and data retrieval speeds.
   *                - `cacheControl`, `contentEncoding`, `contentDisposition`, `contentMd5`: Optional parameters for controlling the caching behavior,
   *                  encoding, content disposition, and integrity check of the blob.
   *
   * Example usage:
   * ```typescript
   * const myBlob = new Blob(this, 'MyBlob', {
   *   name: 'exampleblob',
   *   storageAccountName: 'mystorageaccount',
   *   storageContainerName: 'mycontainer',
   *   type: 'Block',
   *   source: './path/to/file.jpg',
   *   contentType: 'image/jpeg',
   *   metadata: {
   *     author: 'John Doe'
   *   }
   * });
   * ```
   * This class initializes a blob with the specified configurations and handles the uploading of content from the specified source.
   */
  constructor(scope: Construct, id: string, props: StorageBlobConfig) {
    super(scope, id);

    // Create a storage container
    new StorageBlob(this, "blob", {
      name: props.name,
      storageAccountName: props.storageAccountName,
      storageContainerName: props.storageContainerName,
      type: props.type,
      source: props.source,
      contentType: props.contentType,
      metadata: props.metadata,
    });

    this.name = props.name;
  }
}
