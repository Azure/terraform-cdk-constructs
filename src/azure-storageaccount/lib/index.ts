export * from './account';
export * from './container';
export * from './fileshare';
export * from './queue';
export * from './table';
