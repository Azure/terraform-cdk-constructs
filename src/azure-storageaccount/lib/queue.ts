import {
  StorageQueue,
  StorageQueueConfig,
} from "@cdktf/provider-azurerm/lib/storage-queue";
import { Construct } from "constructs";

export class Queue extends Construct {
  public readonly name: string;

  /**
   * Represents an Azure Storage Queue within a specific Azure Storage Account.
   *
   * This class is responsible for the creation and management of an Azure Storage Queue, which is a service for storing large numbers
   * of messages that can be accessed from anywhere in the world via authenticated calls using HTTP or HTTPS. A single queue message
   * can be up to 64 KB in size, and a queue can contain millions of messages, up to the total capacity limit of a storage account.
   * This class provides a way to manage messages in a scalable and secure manner.
   *
   * @param scope - The scope in which to define this construct, typically representing the Cloud Development Kit (CDK) stack.
   * @param id - The unique identifier for this instance of the queue.
   * @param props - Configuration properties for the Azure Storage Queue. These properties may include:
   *                - `name`: The name of the queue, which must be unique within the storage account.
   *                - `storageAccountName`: The name of the storage account in which this queue is being created.
   *                - `metadata`: A dictionary of strings that represents metadata to associate with the queue.
   *                - `timeouts`: Custom timeout settings for CRUD operations on the queue to manage operation durations and retries.
   *
   * Example usage:
   * ```typescript
   * const myQueue = new Queue(this, 'MyQueue', {
   *   name: 'taskqueue',
   *   storageAccountName: 'mystorageaccount',
   *   metadata: {
   *     department: 'IT'
   *   }
   * });
   * ```
   * This class initializes a storage queue with the specified configurations. It can be used to enqueue and process messages as needed
   * within applications, providing a reliable messaging solution for asynchronous communication and coordination of tasks across systems.
   */
  constructor(scope: Construct, id: string, props: StorageQueueConfig) {
    super(scope, id);

    // Create a storage container
    const newQueue = new StorageQueue(this, "Queue", {
      name: props.name,
      storageAccountName: props.storageAccountName,
      metadata: props.metadata,
    });

    this.name = newQueue.name;
  }
}
