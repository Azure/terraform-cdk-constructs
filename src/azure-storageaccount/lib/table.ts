import {
  StorageTable,
  StorageTableConfig,
} from "@cdktf/provider-azurerm/lib/storage-table";
import { Construct } from "constructs";

export class Table extends Construct {
  public readonly name: string;

  constructor(scope: Construct, id: string, props: StorageTableConfig) {
    super(scope, id);

    // Create a storage container
    const table = new StorageTable(this, "table", {
      name: props.name,
      storageAccountName: props.storageAccountName,
      acl: props.acl,
    });

    this.name = table.name;
  }
}
