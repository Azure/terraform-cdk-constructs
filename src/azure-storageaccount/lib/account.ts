import { ResourceGroup } from "@cdktf/provider-azurerm/lib/resource-group";
import { StorageAccount } from "@cdktf/provider-azurerm/lib/storage-account";
import {
  StorageAccountNetworkRulesA,
  StorageAccountNetworkRulesPrivateLinkAccessA,
} from "@cdktf/provider-azurerm/lib/storage-account-network-rules";
import { StorageTableAcl } from "@cdktf/provider-azurerm/lib/storage-table";
import { Construct } from "constructs";
import { Container } from "./container";
import { FileShare, FileShareProps } from "./fileshare";
import { Queue } from "./queue";
import { Table } from "./table";
import { AzureResourceWithAlert } from "../../core-azure/lib";

export interface NetworkRulesProps {
  /**
   * Specifies which traffic to bypass from the network rules. The possible values are 'AzureServices', 'Logging', 'Metrics',
   * and 'None'. Bypassing 'AzureServices' enables Azure's internal services to access the storage account.
   */
  readonly bypass?: string[];

  /**
   * The default action of the network rule set. Options are 'Allow' or 'Deny'. Set to 'Deny' to enable network rules and restrict
   * access to the storage account. 'Allow' permits access by default.
   */
  readonly defaultAction: string;

  /**
   * An array of IP rules to allow access to the storage account. These are specified as CIDR ranges.
   * Example: ['1.2.3.4/32', '5.6.7.0/24'] to allow specific IPs/subnets.
   */
  readonly ipRules?: string[];

  /**
   * An array of virtual network subnet IDs that are allowed to access the storage account. This enables you to secure the storage
   * account to a specific virtual network and subnet within Azure.
   */
  readonly virtualNetworkSubnetIds?: string[];

  /**
   * An array of objects representing the private link access settings. Each object in the array defines the sub-resource name
   * (e.g., 'blob', 'file') and its respective private endpoint connections for the storage account.
   */
  readonly privateLinkAccess?: StorageAccountNetworkRulesPrivateLinkAccessA[];
}

export interface AccountProps {
  /**
   * The type of replication to use for the storage account. This determines how your data is replicated across Azure's infrastructure.
   * Example values: LRS (Locally Redundant Storage), GRS (Geo-Redundant Storage), RAGRS (Read Access Geo-Redundant Storage).
   */
  readonly accountReplicationType?: string;

  /**
   * The performance tier of the storage account. Determines the type of hardware and performance level.
   * Example values: Standard, Premium.
   */
  readonly accountTier?: string;

  /**
   * The Azure region in which to create the storage account.
   */
  readonly location: string;

  /**
   * The name of the storage account. Must be unique across Azure.
   */
  readonly name: string;

  /**
   * The name of the Azure resource group in which to create the storage account.
   */
  readonly resourceGroup?: ResourceGroup;

  /**
   * Tags to apply to the storage account, used for categorization and billing purposes.
   * Format: { [key: string]: string }
   */
  readonly tags?: { readonly [key: string]: string };

  /**
   * A boolean flag indicating whether to enforce HTTPS for data transfer to the storage account.
   */
  readonly enableHttpsTrafficOnly?: boolean;

  /**
   * Managed Service Identity (MSI) details. Used for enabling and managing Azure Active Directory (AAD) authentication.
   */
  readonly identity?: any; // Replace 'any' with a more specific type if available.

  /**
   * The data access tier of the storage account, which impacts storage costs and data retrieval speeds.
   * Example values: Hot, Cool.
   */
  readonly accessTier?: string;

  /**
   * A flag indicating whether the Hierarchical Namespace (HNS) is enabled, which is required for Azure Data Lake Storage Gen2 features.
   */
  readonly isHnsEnabled?: boolean;

  /**
   * The minimum TLS version to be used for securing connections to the storage account.
   * Example values: TLS1_0, TLS1_1, TLS1_2.
   */
  readonly minTlsVersion?: string;

  /**
   * A boolean flag indicating whether public network access to the storage account is allowed.
   */
  readonly publicNetworkAccessEnabled?: boolean;

  /**
   * Docs at Terraform Registry: {@link https://registry.terraform.io/providers/hashicorp/azurerm/3.92.0/docs/resources/storage_account#account_kind StorageAccount#account_kind}
   */
  readonly accountKind?: string;
}

export class Account extends AzureResourceWithAlert {
  public readonly props: AccountProps;
  public id: string;
  public readonly name: string;
  public readonly location: string;
  public resourceGroup: ResourceGroup;
  public readonly accountKind: string;
  public readonly accountTier: string;
  private readonly containers: Map<string, Container>;
  private readonly shares: Map<string, FileShare>;
  private readonly tables: Map<string, Table>;

  /**
   * Represents an Azure Storage Account within a Terraform deployment.
   *
   * This class is responsible for the creation and management of an Azure Storage Account, which is a scalable and secure service
   * for storing large amounts of unstructured data that can be accessed from anywhere in the world over HTTP or HTTPS. Common uses
   * of the Azure Storage Account include storing of blobs (objects), file shares, tables, and queues. This class provides methods
   * to manage storage resources, configure network rules, and integrate with Azure Active Directory for secure access management.
   *
   * @param scope - The scope in which to define this construct, typically representing the Cloud Development Kit (CDK) stack.
   * @param id - The unique identifier for this instance of the storage account.
   * @param props - The properties required to configure the Azure Storage Account, as defined in the AccountProps interface. These include:
   *                - `name`: The name of the storage account, which must be unique within the Azure region.
   *                - `location`: The Azure region where the storage account will be created.
   *                - `resourceGroup`: The Azure Resource Group under which the storage account will be deployed.
   *                - `accountReplicationType`: The type of data replication to ensure data durability and availability.
   *                - `accountTier`: The performance tier that affects the type of hardware used for the storage account.
   *                - `tags`: A dictionary of tags to apply to the storage account for organizational purposes.
   *
   * Example usage:
   * ```typescript
   * const storageAccount = new Account(this, 'MyStorageAccount', {
   *   location: 'East US',
   *   name: 'myStorageAccount',
   *   resourceGroup: myResourceGroup,
   *   accountReplicationType: 'LRS',
   *   accountTier: 'Standard',
   *   enableHttpsTrafficOnly: true,
   *   tags: {
   *     environment: 'production'
   *   }
   * });
   * ```
   * This class sets up the storage account with the specified configurations, handles resource allocation, and applies security
   * settings based on the properties provided.
   */
  constructor(scope: Construct, id: string, props: AccountProps) {
    super(scope, id);

    this.props = props;
    this.resourceGroup = this.setupResourceGroup(props);
    this.containers = new Map<string, Container>();
    this.shares = new Map<string, FileShare>();
    this.tables = new Map<string, Table>();
    this.name = props.name;
    this.location = props.location;

    // default Storage Account Settings
    const defaults = {
      accountReplicationType: "LRS",
      accountTier: "Standard",
      enableHttpsTrafficOnly: true,
      accessTier: "Hot",
      isHnsEnabled: true,
      minTlsVersion: "TLS1_2",
      publicnetworkAccessEnabled: false,
      ...props,
    };

    // Create the storage account
    const storageAccount = new StorageAccount(this, "storageaccount", {
      ...defaults,
      name: props.name,
      resourceGroupName: this.resourceGroup.name,
      location: props.location,
      tags: props.tags,
    });

    this.id = storageAccount.id;
    this.name = storageAccount.name;
    this.accountKind = storageAccount.accountKind;
    this.accountTier = storageAccount.accountTier;
  }

  /**
   * Adds a new container to the storage account.
   * @param name The name of the container. It must be unique within the storage account.
   * @param containerAccessType The level of public access to the container. Defaults to 'private'.
   * @param metadata Optional metadata for the container as key-value pairs.
   * @returns The created Container instance.
   * @throws Error if a container with the same name already exists within the storage account.
   *
   * This method creates a new container within the Azure storage account, allowing for the specification of access
   * level and metadata. If the container already exists, it throws an error to prevent duplication.
   *
   * Example usage:
   * ```typescript
   * const container = storageAccount.addContainer('myContainer', 'private', { owner: 'IT' });
   * ```
   */

  public addContainer(
    name: string,
    containerAccessType?: string,
    metadata?: { [key: string]: string },
  ): Container {
    if (this.containers.has(name)) {
      throw new Error(`Container '${name}' already exists.`);
    }

    const newContainer = new Container(this, name, {
      name: name,
      storageAccountName: this.name,
      containerAccessType: containerAccessType || "private",
      metadata: metadata || {},
    });

    this.containers.set(name, newContainer);
    return newContainer;
  }

  /**
   * Adds a new file share to the storage account.
   * @param name The name of the file share. Must be unique within the storage account.
   * @param props Optional properties for configuring the file share, such as quota and access tier.
   * @returns The created FileShare instance.
   * @throws Error if a file share with the same name already exists.
   *
   * This method facilitates the addition of a file share to the storage account, with optional settings for
   * capacity (quota) and data access frequency (access tier). If a file share with the same name exists, an error is thrown.
   *
   * Example usage:
   * ```typescript
   * const fileShare = storageAccount.addFileShare('myFileShare', { quota: 1024, accessTier: 'Hot' });
   * ```
   */

  public addFileShare(name: string, props?: FileShareProps): FileShare {
    if (this.shares.has(name)) {
      throw new Error(`Share '${name}' already exists.`);
    }

    const defaults = {
      quota: props?.quota || 1024,
      accessTier: props?.accessTier || "Hot",
      enabledProtocol: props?.enabledProtocol || "SMB",
      acl: props?.acl || [],
      metadata: props?.metadata || {},
    };

    const newShare = new FileShare(this, name, {
      ...defaults,
      name: name,
      storageAccountName: this.name,
    });

    this.shares.set(name, newShare);
    return newShare;
  }

  /**
   * Adds a new table to the storage account.
   * @param name The name of the table. Must be unique within the storage account.
   * @param acl Optional access control list for the table, specifying permissions.
   * @returns The created Table instance.
   * @throws Error if a table with the same name already exists.
   *
   * This method creates a new table within the storage account, optionally allowing for access control configurations.
   * It throws an error if a table with the same name already exists, ensuring uniqueness within the account.
   *
   * Example usage:
   * ```typescript
   * const table = storageAccount.addTable('myTable', [{ id: 'policy1', type: 'read' }]);
   * ```
   */

  public addTable(name: string, acl?: StorageTableAcl[]): Table {
    if (this.tables.has(name)) {
      throw new Error(`Table '${name}' already exists.`);
    }

    const newTable = new Table(this, name, {
      name: name,
      storageAccountName: this.name,
      acl: acl,
    });

    this.tables.set(name, newTable);
    return newTable;
  }

  /**
   * Adds a new queue to the storage account.
   * @param name The name of the queue. Must be unique within the storage account.
   * @param metadata Optional metadata for the queue as key-value pairs.
   * @returns The created Queue instance.
   *
   * This method creates a new queue in the storage account, with optional metadata. It is useful for message queuing
   * in applications, enabling asynchronous task processing and inter-service communication.
   *
   * Example usage:
   * ```typescript
   * const queue = storageAccount.addQueue('myQueue', { priority: 'high' });
   * ```
   */

  public addQueue(name: string, metadata?: { [key: string]: string }): Queue {
    return new Queue(this, name, {
      name: name,
      storageAccountName: this.name,
      metadata: metadata,
    });
  }

  /**
   * Adds network rules to the storage account to control access based on IP and virtual network settings.
   * @param props Configuration properties for the network rules, including allowed IPs and virtual network subnet IDs.
   * @returns The configured network rules.
   *
   * This method configures network rules for the storage account, specifying which IPs and virtual networks can access
   * the storage resources. It allows detailed control over data security and access management.
   *
   * Example usage:
   * ```typescript
   * storageAccount.addNetworkRules({
   *   bypass: ['AzureServices'],
   *   defaultAction: 'Deny',
   *   ipRules: ['1.2.3.4/32'],
   *   virtualNetworkSubnetIds: ['subnetId'],
   * });
   * ```
   */

  public addNetworkRules(
    props: NetworkRulesProps,
  ): StorageAccountNetworkRulesA {
    return new StorageAccountNetworkRulesA(this, "rules", {
      storageAccountId: this.id,
      bypass: props.bypass,
      defaultAction: props.defaultAction,
      ipRules: props.ipRules,
      virtualNetworkSubnetIds: props.virtualNetworkSubnetIds,
      privateLinkAccess: props.privateLinkAccess,
    });
  }
}
