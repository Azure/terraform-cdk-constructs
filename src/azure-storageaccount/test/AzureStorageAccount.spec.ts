import { Testing, TerraformStack} from 'cdktf';
import 'cdktf/lib/testing/adapters/jest';
import * as storage from "..";
import {AzurermProvider} from "@cdktf/provider-azurerm/lib/provider";
import { exampleAzureStorageAccount } from '../test/ExampleAzureStorageAccount';

describe('Azure Storage Account With Defaults', () => {
  let stack: TerraformStack;
  let fullSynthResult: any;

  beforeEach(() => {
    const app = Testing.app();
    stack = new TerraformStack(app, "test");

    new AzurermProvider(stack, "azureFeature", {features: {}});

    // Create a Storage Account with the defined rules
    new storage.Account(stack, "storageaccount", {
      name: "test42348808",
      location: "eastus",
    });

    fullSynthResult = Testing.fullSynth(stack); // Save the result for reuse
  });

  it("renders an Azure Storage Account with defaults and checks snapshot", () => {
    expect(
      Testing.synth(stack)
    ).toMatchSnapshot(); // Compare the already prepared stack
  });

  it("check if the produced terraform configuration is valid", () => {
    expect(fullSynthResult).toBeValidTerraform(); // Use the saved result
  });

  it("check if this can be planned", () => {
    expect(fullSynthResult).toPlanSuccessfully(); // Use the saved result
  });
});


describe('Azure Storage Account Example', () => {
  
  it("renders the Example Storage Account and checks snapshot", () => {
  
    expect(
      Testing.synth(new exampleAzureStorageAccount(Testing.app(), "testNetworkSecurityGroupExample"))
      
    ).toMatchSnapshot();
});

  it("check if the produced terraform configuration is valid", () => {
    // We need to do a full synth to plan the terraform configuration
    expect(Testing.fullSynth(new exampleAzureStorageAccount(Testing.app(), "testNetworkSecurityGroupExample"))).toBeValidTerraform();
  });

  it("check if this can be planned", () => {
  
    
    // We need to do a full synth to plan the terraform configuration
    expect(Testing.fullSynth(new exampleAzureStorageAccount(Testing.app(), "testNetworkSecurityGroupExample"))).toPlanSuccessfully();
  });
});