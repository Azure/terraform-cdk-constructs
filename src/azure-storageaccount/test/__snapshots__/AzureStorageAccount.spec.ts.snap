// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Azure Storage Account Example renders the Example Storage Account and checks snapshot 1`] = `
"{
  "data": {
    "azurerm_client_config": {
      "CurrentClientConfig": {
      }
    },
    "azurerm_monitor_diagnostic_categories": {
      "storageaccount_diagsettings_diagcategories_FDFFAC42": {
        "resource_id": "\${azurerm_storage_account.storageaccount_5B90F750.id}"
      }
    }
  },
  "output": {
    "resource_group_name": {
      "value": "\${azurerm_resource_group.rg.name}"
    },
    "storage_account_account_kind": {
      "value": "\${azurerm_storage_account.storageaccount_5B90F750.account_kind}"
    },
    "storage_account_account_tier": {
      "value": "\${azurerm_storage_account.storageaccount_5B90F750.account_tier}"
    },
    "storage_account_name": {
      "value": "\${azurerm_storage_account.storageaccount_5B90F750.name}"
    },
    "storage_container_name": {
      "value": "testcontainer"
    }
  },
  "provider": {
    "azurerm": [
      {
        "features": {
        }
      }
    ]
  },
  "resource": {
    "azurerm_log_analytics_workspace": {
      "log_analytics": {
        "location": "eastus",
        "name": "la-\${var.name}",
        "resource_group_name": "\${azurerm_resource_group.rg.name}"
      }
    },
    "azurerm_monitor_diagnostic_setting": {
      "storageaccount_diagsettings_1F202998": {
        "dynamic": {
          "enabled_log": {
            "content": {
              "category": "\${enabled_log.value}"
            },
            "for_each": "\${tolist(data.azurerm_monitor_diagnostic_categories.storageaccount_diagsettings_diagcategories_FDFFAC42.log_category_types)}"
          },
          "metric": {
            "content": {
              "category": "\${metric.value}"
            },
            "for_each": "\${tolist(data.azurerm_monitor_diagnostic_categories.storageaccount_diagsettings_diagcategories_FDFFAC42.metrics)}"
          }
        },
        "log_analytics_workspace_id": "\${azurerm_log_analytics_workspace.log_analytics.id}",
        "name": "diagsettings",
        "target_resource_id": "\${azurerm_storage_account.storageaccount_5B90F750.id}"
      }
    },
    "azurerm_resource_group": {
      "rg": {
        "location": "eastus",
        "name": "rg-\${var.name}"
      }
    },
    "azurerm_role_assignment": {
      "storageaccount_TfTokenTOKEN50Contributor_role_CF7C7E2F": {
        "principal_id": "\${data.azurerm_client_config.CurrentClientConfig.object_id}",
        "role_definition_name": "Contributor",
        "scope": "\${azurerm_storage_account.storageaccount_5B90F750.id}"
      }
    },
    "azurerm_storage_account": {
      "storageaccount_5B90F750": {
        "access_tier": "Hot",
        "account_replication_type": "LRS",
        "account_tier": "Standard",
        "enable_https_traffic_only": true,
        "is_hns_enabled": true,
        "location": "eastus",
        "min_tls_version": "TLS1_2",
        "name": "sta\${var.name}8898",
        "resource_group_name": "\${azurerm_resource_group.rg.name}"
      }
    },
    "azurerm_storage_account_network_rules": {
      "storageaccount_rules_4094FBC9": {
        "bypass": [
          "AzureServices"
        ],
        "default_action": "Deny",
        "ip_rules": [
          "0.0.0.0/0"
        ],
        "storage_account_id": "\${azurerm_storage_account.storageaccount_5B90F750.id}"
      }
    },
    "azurerm_storage_blob": {
      "storageaccount_testcontainer_testblobtxt_blob_0EC18225": {
        "content_type": "application/octet-stream",
        "metadata": {
        },
        "name": "testblob.txt",
        "source": "../../../test.txt",
        "storage_account_name": "\${azurerm_storage_account.storageaccount_5B90F750.name}",
        "storage_container_name": "\${azurerm_storage_container.storageaccount_testcontainer_D98810FC.name}",
        "type": "Block"
      }
    },
    "azurerm_storage_container": {
      "storageaccount_testcontainer2_container_7DE9A01E": {
        "container_access_type": "private",
        "metadata": {
        },
        "name": "testcontainer2",
        "storage_account_name": "\${azurerm_storage_account.storageaccount_5B90F750.name}"
      },
      "storageaccount_testcontainer_D98810FC": {
        "container_access_type": "private",
        "metadata": {
        },
        "name": "testcontainer",
        "storage_account_name": "\${azurerm_storage_account.storageaccount_5B90F750.name}"
      }
    },
    "azurerm_storage_queue": {
      "storageaccount_testqueue_Queue_66867FEF": {
        "name": "testqueue",
        "storage_account_name": "\${azurerm_storage_account.storageaccount_5B90F750.name}"
      }
    },
    "azurerm_storage_share": {
      "storageaccount_testshare_F365FFBE": {
        "access_tier": "Hot",
        "acl": [
        ],
        "enabled_protocol": "SMB",
        "metadata": {
        },
        "name": "testshare",
        "quota": 1024,
        "storage_account_name": "\${azurerm_storage_account.storageaccount_5B90F750.name}"
      }
    },
    "azurerm_storage_share_file": {
      "storageaccount_testshare_file_BF2AC943": {
        "content_type": "application/octet-stream",
        "metadata": {
        },
        "name": "testfile.txt",
        "source": "../../../test.txt",
        "storage_share_id": "\${azurerm_storage_share.storageaccount_testshare_F365FFBE.id}"
      }
    },
    "azurerm_storage_table": {
      "storageaccount_testtable_4353D645": {
        "name": "testtable",
        "storage_account_name": "\${azurerm_storage_account.storageaccount_5B90F750.name}"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.83.0"
      }
    }
  },
  "variable": {
    "name": {
      "default": "test",
      "description": "System name used to randomize the resources",
      "type": "string"
    }
  }
}"
`;

exports[`Azure Storage Account With Defaults renders an Azure Storage Account with defaults and checks snapshot 1`] = `
"{
  "provider": {
    "azurerm": [
      {
        "features": {
        }
      }
    ]
  },
  "resource": {
    "azurerm_resource_group": {
      "storageaccount_rg_F4452564": {
        "location": "eastus",
        "name": "rg-test42348808"
      }
    },
    "azurerm_storage_account": {
      "storageaccount_5B90F750": {
        "access_tier": "Hot",
        "account_replication_type": "LRS",
        "account_tier": "Standard",
        "enable_https_traffic_only": true,
        "is_hns_enabled": true,
        "location": "eastus",
        "min_tls_version": "TLS1_2",
        "name": "test42348808",
        "resource_group_name": "\${azurerm_resource_group.storageaccount_rg_F4452564.name}"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.83.0"
      }
    }
  }
}"
`;
